superset/assets/javascripts/chart/Chart.jsx

```diff
@@ -0,0 +1,190 @@
+    this.removeFilter = this.removeFilter.bind(this);
+  }
+
+  componentDidMount() {
+    this.runQuery();
+  }
+
+  componentWillReceiveProps(nextProps) {
+    this.containerId = nextProps.containerId;
+    this.selector = `#${this.containerId}`;
+    this.formData = nextProps.formData;
+    this.datasource = nextProps.datasource;
+  }
+
+  componentDidUpdate(prevProps) {
+    if (
+      this.props.queryResponse &&
+      this.props.chartStatus === 'success' &&
+      !this.props.queryResponse.error && (
+      prevProps.queryResponse !== this.props.queryResponse ||
+      prevProps.height !== this.props.height ||
+      prevProps.width !== this.props.width ||
+      this.props.triggerRender)
+    ) {
+      this.renderViz();
+    }
+  }
+
+  getFilters() {
+    return this.props.getFilters();
+  }
+
+  addFilter(col, vals, merge = true, refresh = true) {
+    this.props.addFilter(col, vals, merge, refresh);
+  }
+
+  clearFilter() {
+    this.props.clearFilter();
+  }
+
+  removeFilter(col, vals) {
+    this.props.removeFilter(col, vals);
+  }
+
+  clearError() {
+    this.setState({
+      errorMsg: null,
+    });
+  }
+
+  width() {
+    return this.props.width ?
+      this.props.width :
+      this.container.el.offsetWidth;
+  }
+
+  height() {
+    return this.props.height ?
+      this.props.height :
+      this.container.el.offsetHeight;
+  }
+
+  d3format(col, number) {
+    const format =
+      this.props.datasource.column_formats && this.props.datasource.column_formats[col] ?
+      this.props.datasource.column_formats[col] : '0.3s';
+    return d3format(format, number);
+  }
+
+  runQuery() {
+    this.props.actions.runQuery(this.props.formData, true,
+      this.props.timeout,
+      this.props.chartKey,
+    );
+  }
+
+  render_template(s) {
+    const context = {
+      width: this.width(),
+      height: this.height(),
+    };
+    return Mustache.render(s, context);
+  }
+
+  renderViz() {
+    this.props.actions.renderTriggered(this.props.chartKey);
+
+    const viz = visMap[this.props.viz_type];
+    try {
+      viz(this, this.props.queryResponse, this.props.actions.setControlValue);
+    } catch (e) {
+      this.props.actions.chartRenderingFailed(e, this.props.chartKey);
+    }
+  }
+
+  render() {
+    const isLoading = this.props.chartStatus === 'loading';
+    return (
+      <div className={`token col-md-12 ${isLoading ? 'is-loading' : ''}`}>
+        {isLoading && <img
+          alt="loading"
+          width="25"
+          src="/static/assets/images/loading.gif"
+          style={{ position: 'absolute' }}
+        />}
+
+        {this.props.chartAlert &&
+        <StackTraceMessage
+          message={this.props.chartAlert}
+          queryResponse={this.props.queryResponse}
+        />
+        }
+
+        {!this.props.chartAlert &&
+        <ChartBody
+          containerId={this.containerId}
+          vizType={this.props.formData.viz_type}
+          height={this.height.bind(this)}
```

Comment:
why do you bind this in the render vs all the other methods in the constructor? you shouldn't bind in a render because it defines a new func every render